### PR TITLE
[WIP] EntityViewService optimization using local cache and lifecycle event broadcasting (declined)

### PR DESCRIPTION
**Pull request by smatvienko-tb**
EntityViewService optimization using local cache and lifecycle event broadcasting

## Pull Request description

During the cluster performance test, the Redis cache reads too frequently during telemetry save and notify because of the entity view update feature.

The aim is to keep the in-memory entity view list owned by devices that are frequently written telemetry.

These are especially useful for the clusters that do not use the entity view feature at all.

The memory map is declared as a soft referenced map, so the only free memory will be used for this optimization.

In the worst-case scenario with no free memory available, the feature will perform calling the Redis cache as before this fix.

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: msa/tb/docker-cassandra/Dockerfile [hunk 1]
```diff
@@ -44,15 +44,17 @@ ENV CASSANDRA_LOG=/var/log/cassandra
 COPY logback.xml ${pkg.name}.conf start-db.sh stop-db.sh start-tb.sh upgrade-tb.sh install-tb.sh ${pkg.name}.deb /tmp/
 
 RUN apt-get update \
-    && apt-get install -y curl nmap procps gnupg2 \
+    && apt-get install -y --no-install-recommends curl nmap procps gnupg2 \
```
Comment (ashvayka) [inline msa/tb/docker-cassandra/Dockerfile:47]: Not related to this PR. Remove this and create new one.

File: pom.xml [hunk 2]
```diff
@@ -98,8 +98,9 @@
         <elasticsearch.version>5.0.2</elasticsearch.version>
         <delight-nashorn-sandbox.version>0.2.1</delight-nashorn-sandbox.version>
         <!-- IMPORTANT: If you change the version of the kafka client, make sure to synchronize our overwritten implementation of the
-        org.apache.kafka.common.network.NetworkReceive class in the application module. It addresses the issue https://issues.apache.org/jira/browse/KAFKA-4090. -->
-        <kafka.version>2.8.0</kafka.version>
+        org.apache.kafka.common.network.NetworkReceive class in the application module. It addresses the issue https://issues.apache.org/jira/browse/KAFKA-4090.
```
Comment (ashvayka) [inline pom.xml:101]: Are you sure the NetworkReceive class was not changed?

File: pom.xml [hunk 3]
```diff
@@ -98,8 +98,9 @@
...
+        Here is the source to track https://github.com/apache/kafka/tree/trunk/clients/src/main/java/org/apache/kafka/common/network -->
+        <kafka.version>3.2.0</kafka.version>
```
Comment (ashvayka) [inline pom.xml:103]: Not related to this PR

File: application/src/main/java/org/thingsboard/server/service/entitiy/entityView/DefaultTbEntityViewService.java [hunk 4]
```diff
@@ -214,6 +227,35 @@ public EntityView unassignEntityViewFromCustomer(TenantId tenantId, EntityViewId
...
+        ListenableFuture<List<EntityView>> future = entityViewService.findEntityViewsByTenantIdAndEntityIdAsync(tenantId, entityId);
+
+        return Futures.transform(future, (entityViewList) -> {
+            localCacheByTenant.put(entityId, entityViewList);
+            return entityViewList;
+        }, MoreExecutors.directExecutor());
+    }
+
+    @Override
+    public void onComponentLifecycleMsg(ComponentLifecycleMsg componentLifecycleMsg) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/entitiy/entityView/DefaultTbEntityViewService.java:247]: please replace with onEntityViewUpdate

File: msa/js-executor/queue/kafkaTemplate.js [hunk 5]
```diff
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-const {logLevel, Kafka, CompressionTypes} = require('kafkajs');
+const {logLevel, Kafka, CompressionTypes, Partitioners} = require('kafkajs');
```
[the remark this answers is not in the document]
Comment (smatvienko-tb) [reply msa/js-executor/queue/kafkaTemplate.js:16]: extracted to PR https://github.com/thingsboard/thingsboard/pull/6782

File: pom.xml (hunk 2 above)
Comment (smatvienko-tb) [reply pom.xml:101]: absolutely sure. It has not been modified for 3 years.
image

File: pom.xml (hunk 3 above)
Comment (smatvienko-tb) [reply pom.xml:103]: extracted to https://github.com/thingsboard/thingsboard/pull/6783

File: common/cache/src/main/java/org/thingsboard/server/cache/RedisTbTransactionalCache.java [hunk 6]
```diff
@@ -47,7 +47,7 @@
     @Getter
     private final String cacheName;
     private final JedisConnectionFactory connectionFactory;
-    private final RedisSerializer<String> keySerializer = new StringRedisSerializer();
+    private final RedisSerializer<String> keySerializer = StringRedisSerializer.UTF_8;
```
[the remark this answers is not in the document]
Comment (smatvienko-tb) [reply common/cache/src/main/java/org/thingsboard/server/cache/RedisTbTransactionalCache.java]: removed

File: application/src/main/java/org/thingsboard/server/service/entitiy/entityView/DefaultTbEntityViewService.java (hunk 4 above)
Comment (smatvienko-tb) [reply application/src/main/java/org/thingsboard/server/service/entitiy/entityView/DefaultTbEntityViewService.java:247]: I would like to set a generic Event sourcing interface for any entity. The idea is to not stick to a particular entity name. I agree that refactoring is required.

This extracted to separate PR https://github.com/thingsboard/thingsboard/pull/6785

**Comment by smatvienko-tb** [comment]
This PR is closed as it splitter to many PRs.
Please, follow the main PR https://github.com/thingsboard/thingsboard/pull/6785

**Pull request closed without merging**